File: problems/chapter15/miiitomi/002.cpp
// https://atcoder.jp/contests/jag2012autumn/tasks/icpc2012autumn_c

#include <bits/stdc++.h>
using namespace std;

struct UnionFind {
    vector<int> par, siz;

    UnionFind(int n) : par(n, -1), siz(n, 1) {}

    int root(int x) {
        if (par[x] == -1) return x;
        else return par[x] = root(par[x]);
    }

    bool issame(int x, int y) {
        return root(x) == root(y);
    }

    bool unite(int x, int y) {
        x = root(x);
        y = root(y);

        if (x == y) return false;

        if (siz[x] < siz[y]) swap(x, y);

        par[y] = x;
        siz[x] += siz[y];
        return true;
    }

    int size(int x) {
        return siz[root(x)];
    }
};

int main() {
    while (true) {
        int n, m;
        cin >> n >> m;
        if (n == 0 && m == 0) return 0;

        vector<pair<int, pair<int, int>>> V(m);
        for (int i = 0; i < m; i++) {
            int s, t, c;
            cin >> s >> t >> c;
            s--;
            t--;
            V[i].first = c;
            V[i].second = make_pair(s, t);
        }
        sort(V.begin(), V.end());

        UnionFind uf(n);

        vector<int> E;
        for (int i = 0; i < m; i++) {
            if (uf.issame(V[i].second.first, V[i].second.second)) continue;
            E.push_back(V[i].first);
            uf.unite(V[i].second.first, V[i].second.second);
        }

        cout << E[n/2 - 1] << endl;
    }
}

File: problems/chapter15/mshibatatt/002.py
# https://atcoder.jp/contests/jag2012autumn/tasks/icpc2012autumn_c?lang=ja
class UnionFind():
    def __init__(self, n) -> None:
        self.par = [-1] * n
        self.siz = [1] * n
    
    # return root of x
    def root(self, x):
        if self.par[x] == -1:
            return x
        else: 
            self.par[x] = self.root(self.par[x])
            return self.par[x]
    
    # return if x and y are in the same group
    def issame(self, x, y):
        return self.root(x) == self.root(y)
    
    # combine group with x and one with y
    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)

        if x == y:
            return False
        if self.siz[x] < self.siz[y]:
            x, y = y, x
        
        # make y child of x
        self.par[y] = x
        self.siz[x] += self.siz[y]

        return True
        
    def size(self, x):
        return self.siz[self.root(x)]


def main():
    n, m = map(int, input().split())
    if n == 0 and m == 0:
        return 'end'
    edge = []
    for _ in range(m):
        s, t, c = map(int, input().split())
        edge.append([s-1, t-1, c])

    edge.sort(key=lambda x: x[2])
    uf = UnionFind(n)
    med = n // 2 
    counter = 0

    for s, t, c in edge:
        if uf.issame(s, t):
            continue
        uf.unite(s, t)
        counter += 1
        if counter == med:
            print(c)
            break
    return 'continue'
    
if __name__=='__main__':
    end_flag = 'continue'
    while end_flag == 'continue':
        end_flag = main()